plugins/module_utils/azure_rm_common.py

```diff
@@ -1584,9 +1584,27 @@ def _get_profile(self, profile="default"):
 
         return None
 
-    def _get_msi_credentials(self, subscription_id=None, client_id=None, **kwargs):
-        credentials = MSIAuthentication(client_id=client_id)
-        credential = MSIAuthenticationWrapper(client_id=client_id)
+    def _get_msi_credentials(self, subscription_id=None, client_id=None, _cloud_environment=None, **kwargs):
+        if not _cloud_environment:
+            cloud_environment = azure_cloud.AZURE_PUBLIC_CLOUD
+        else:
+            # try to look up "well-known" values via the name attribute on azure_cloud members
+            all_clouds = [x[1] for x in inspect.getmembers(azure_cloud) if isinstance(x[1], azure_cloud.Cloud)]
+            matched_clouds = [x for x in all_clouds if x.name == _cloud_environment]
+            if len(matched_clouds) == 1:
+                cloud_environment = matched_clouds[0]
+            elif len(matched_clouds) > 1:
+                self.fail("Azure SDK failure: more than one cloud matched for cloud_environment name '{0}'".format(cloud_environment))
+            else:
+                if not urlparse.urlparse(cloud_environment).scheme:
+                    self.fail("cloud_environment must be an endpoint discovery URL or one of {0}".format([x.name for x in all_clouds]))
+                try:
+                    self._cloud_environment = azure_cloud.get_cloud_from_metadata_endpoint(cloud_environment)
+                except Exception as e:
+                    self.fail("cloud_environment {0} could not be resolved: {1}".format(cloud_environment, e.message), exception=traceback.format_exc())
+
+        credentials = MSIAuthentication(client_id=client_id, cloud_environment=cloud_environment)
+        credential = MSIAuthenticationWrapper(client_id=client_id, cloud_environment=cloud_environment)
```

Comment:
```suggestion
    def _get_msi_credentials(self, subscription_id=None, client_id=None, _cloud_environment="AzureCloud", **kwargs):
        try:
            all_clouds = [x[1].name for x in inspect.getmembers(azure_cloud) if isinstance(x[1], azure_cloud.Cloud)]
            if _cloud_environment not in all_clouds:
                self.fail(f"Azure SDK failure: no cloud_environment name matched {_cloud_environment}")
            self._cloud_environment = azure_cloud.get_cloud_from_metadata_endpoint(_cloud_environment)
        except Exception as e:
            self.fail(f"cloud_environment {_cloud_environment} could not be resolved: {e.message}", exception=traceback.format_exc())
        
        credentials = MSIAuthentication(client_id=client_id, cloud_environment=_cloud_environment)
        credential = MSIAuthenticationWrapper(client_id=client_id, cloud_environment=_cloud_environment)
```

I believe this would be a bit cleaner.

Reply:
overall I think this is the fix we need for https://github.com/ansible-collections/azure/issues/893

Reply:
one more note:
I think we could add a `_get_azure_cloud` function since the `_get_azure_cli_credentials`  could benefit from that as well:

```python
def _get_azure_cloud(self, cloud_environment: str) -> azure_cloud.Cloud:
    all_clouds = {x[1].name:x[1] for x in inspect.getmembers(azure_cloud) if isinstance(x[1], azure_cloud.Cloud)}
    if _cloud_environment not in all_clouds:
        self.fail(f"Azure SDK failure: no cloud_environment name matched {_cloud_environment}")
    return all_clouds[_cloud_environment]
```